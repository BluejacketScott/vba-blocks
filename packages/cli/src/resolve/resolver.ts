import { Config } from '../config';
import { Version, Dependency } from '../manifest';
import { isRegistryDependency } from '../manifest/dependency';
import Manager, { Registration } from '../sources';
import { DependencyGraph } from './dependency-graph';

export interface Resolution {
  name: string;
  range: string[];
  preferred?: Registration;
  registered: Registration[];
}
export type ResolutionGraph = Map<string, Resolution>;

export default class Resolver {
  config: Config;
  manager: Manager;
  graph: ResolutionGraph;
  loading: Map<string, Promise<Registration[]>>;
  preferred: Map<string, Registration>;

  constructor(config: Config) {
    this.manager = new Manager(config);
    this.graph = new Map();
    this.loading = new Map();
    this.preferred = new Map();
  }

  async update() {
    await this.manager.update();
  }

  async get(dependency: Dependency): Promise<Resolution> {
    const { name } = dependency;

    if (this.loading.has(name)) await this.loading.get(name);
    let resolution: Resolution = this.graph.get(name);

    if (!resolution) {
      const loading = this.manager.resolve(dependency);
      this.loading.set(name, loading);

      const registered = await loading;
      resolution = {
        name,
        range: [],
        registered
      };

      if (this.preferred.has(name)) {
        resolution.preferred = this.preferred.get(name);
      }

      this.loading.delete(name);
      this.graph.set(name, resolution);
    }

    if (isRegistryDependency(dependency)) {
      resolution.range.push(dependency.version);
    }

    return resolution;
  }

  getRegistration(id: string) {
    const [name, version] = id.split('@', 2);
    const { registered } = this.graph.get(name);
    return registered.find(registration => registration.version === version);
  }

  prefer(preferred: DependencyGraph) {
    for (const registration of preferred) {
      this.preferred.set(registration.name, registration);
    }
  }

  [Symbol.iterator]() {
    return this.graph.entries();
  }
}
