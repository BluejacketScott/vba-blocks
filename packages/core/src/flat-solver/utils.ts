export type Iterable = any;

export function forEach(
  iterator: Iterable,
  fn: (item: any, index: number) => any
) {
  let index = 0;
  for (const item of iterator) {
    fn(item, index++);
  }
}

export function map(
  iterator: Iterable,
  fn: (item: any, index: number) => any
): any[] {
  let index = 0;
  const results = [];
  for (const item of iterator) {
    results.push(fn(item, index++));
  }
  return results;
}

export function pluck(values: Iterable, key: string): any[] {
  return map(values, value => value && value[key]);
}

export function reduce<T>(
  iterator: Iterable,
  fn: (memo: T, item: any) => T,
  memo: T
) {
  for (const item of iterator) {
    memo = fn(memo, item);
  }

  return memo;
}

export function unique(values: Iterable): any[] {
  const asSet = new Set(values);
  return Array.from(asSet);
}
