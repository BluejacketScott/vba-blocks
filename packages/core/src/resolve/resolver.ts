import { Config } from '../config';
import { Version, Dependency } from '../manifest';
import {
  isRegistryDependency,
  isPathDependency,
  isGitDependency
} from '../manifest/dependency';
import { Registration, registry, path, git } from '../manager';

export interface Resolution {
  name: string;
  range: string[];
  registered: Registration[];
}
export type ResolutionGraph = Map<string, Resolution>;

export default class Resolver {
  config: Config;
  graph: ResolutionGraph;
  loading: Map<string, Promise<Registration[]>>;

  constructor(config: Config) {
    this.graph = new Map();
  }

  async get(dependency: Dependency): Promise<Resolution> {
    const { name } = dependency;

    if (this.loading.has(name)) await this.loading.get(name);
    let resolution = this.graph.get(name);

    if (!resolution) {
      let loading;
      if (isRegistryDependency(dependency)) {
        loading = registry.resolve(this.config, dependency);
      } else if (isPathDependency(dependency)) {
        loading = path.resolve(this.config, dependency);
      } else {
        loading = git.resolve(this.config, dependency);
      }

      this.loading.set(name, loading);

      const registered = await loading;
      const resolution = {
        name,
        range: [],
        registered
      };

      this.loading.delete(name);
      this.graph.set(name, resolution);
    }

    let rangeInfo;
    if (isRegistryDependency(dependency)) {
      rangeInfo = dependency.version;
    } else if (isPathDependency(dependency)) {
      rangeInfo = dependency.path;
    } else {
      const { git, branch, tag, rev } = dependency;
      rangeInfo = { git, branch, tag, rev };
    }
    resolution.range.push(rangeInfo);

    return resolution;
  }

  getRegistration(id: string) {
    const [name, version] = id.split('@');
    const { registered } = this.graph.get(name);
    return registered.find(registration => registration.version === version);
  }

  [Symbol.iterator]() {
    return this.graph.entries();
  }
}
