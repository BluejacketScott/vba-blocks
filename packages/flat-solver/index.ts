export type Dependencies = any;
export interface Resolver {}
export interface Solution {}

export async function solve(
  dependencies: Dependencies,
  resolver: Resolver
): Promise<Solution> {
  return Promise.resolve({});
}
