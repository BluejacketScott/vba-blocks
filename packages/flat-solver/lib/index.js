'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
async function solve(dependencies, resolver) {
  return Promise.resolve({});
}
exports.solve = solve;
