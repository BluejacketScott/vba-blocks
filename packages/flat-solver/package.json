{
  "name": "flat-solver",
  "private": true,
  "version": "0.0.0",
  "repository": "https://github.com/vba-blocks/vba-blocks.git",
  "author": "Tim Hall <tim.hall.engr@gmail.com>",
  "license": "MIT",
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "scripts": {
    "build": "tsc --outDir lib"
  },
  "dependencies": {
    "logic-solver": "^2.0.1",
    "semver": "^5.4.1"
  },
  "devDependencies": {
    "typescript": "^2.4.2"
  }
}
