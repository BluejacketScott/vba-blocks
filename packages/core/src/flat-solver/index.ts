import * as semver from 'semver';
import * as Logic from 'logic-solver';
import { forEach, map, pluck } from './utils';
import {
  resolveDependencies,
  FlatGraph,
  FlatGraphNode
} from './resolve-dependencies';

export { resolveDependencies, FlatGraph, FlatGraphNode };

export type Version = string;
export type Dependency = { name: string; version: Version };

export interface Resolver {
  (name: string): Promise<PackageInfo[]>;
}

export interface PackageInfo {
  id?: string;
  name: string;
  version: Version;
  dependencies: Dependency[];
}

export interface Solution {
  success: boolean;
  result?: string[];
}

/**
 * Solve for the resolved flat dependency tree given a set of root dependencies
 */
export async function solve(
  dependencies: Dependency[] | { [name: string]: Version },
  resolver: Resolver
): Promise<Solution> {
  if (!Array.isArray(dependencies))
    dependencies = toArray<Dependency>(dependencies);

  const graph = await resolveDependencies(dependencies, resolver);
  const solver = new Logic.Solver();

  forEach(graph.values(), node => {
    const { name, versions, required } = node;
    const ids = pluck(versions, 'id');

    if (required) {
      solver.require(Logic.exactlyOne(...ids));
    } else {
      solver.require(Logic.atMostOne(...ids));
    }
  });

  const reversed = [...graph.values()].reverse();
  forEach(reversed, node => {
    const { versions } = node;

    forEach(versions, info => {
      const { id } = info;
      forEach(info.dependencies, dependency => {
        const matching = getMatching(dependency, graph);
        solver.require(Logic.implies(id, Logic.or(...matching)));
      });
    });
  });

  const solution = solver.solve();
  if (!solution) return { success: false };

  return {
    success: true,
    result: solution.getTrueVars()
  };
}

function getMatching(dependency, graph) {
  const { name, version: req } = dependency;
  const versions = graph.get(name).versions;

  const matching = versions
    .filter(({ version }) => semver.satisfies(version, req))
    .map(version => version.id);

  return matching;
}

function toArray<T>(dependencies): T[] {
  return map(Object.entries(dependencies), ([name, version]) => {
    return { name, version };
  });
}
