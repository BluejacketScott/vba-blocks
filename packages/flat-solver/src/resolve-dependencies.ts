import * as semver from 'semver';
import { Version, Dependency, PackageInfo, Resolver } from './';
import { forEach, unique } from './utils';

export type FlatGraph = Map<string, FlatGraphNode>;
export type FlatGraphNode = {
  name: string;
  needs: Version[];
  versions: PackageInfo[];
  required?: boolean;
};

export async function resolveDependencies(
  dependencies: Dependency[],
  resolver: Resolver,
  graph?: FlatGraph
): Promise<FlatGraph> {
  // For top-level dependencies, mark as required
  const required = !graph;
  if (!graph) graph = new Map();

  for (const dependency of dependencies) {
    const { name, version } = dependency;
    if (!graph.has(name)) {
      const versions = await resolver(name);
      const node: FlatGraphNode = {
        name,
        needs: [],
        versions,
        required
      };

      graph.set(name, node);

      for (const version of versions) {
        await resolveDependencies(version.dependencies, resolver, graph);
      }
    }

    graph.get(name).needs.push(version);
  }

  if (required) {
    // Optimize graph by filtering out unneeded versions
    forEach(graph.values(), (node: FlatGraphNode) => {
      const { name, versions } = node;
      const needs: string[] = unique(node.needs);
      const range = needs.join(' || ');

      node.needs = needs;
      node.versions = versions
        .filter(({ version }) => semver.satisfies(version, range))
        .map(info => Object.assign({ id: `${name}@${info.version}` }, info))
        .reverse();
    });
  }

  return graph;
}
