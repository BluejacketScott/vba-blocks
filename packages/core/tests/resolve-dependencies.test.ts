import { resolveDependencies } from '../src/flat-solver/resolve-dependencies';
import resolver from './helpers/resolver';

test('resolves simple dependency tree', async () => {
  const dependencies = [{ name: 'a', version: '^1.0.0' }];
  const flatGraph = await resolveDependencies(dependencies, resolver);

  expect(flatGraph).toMatchSnapshot();
});
