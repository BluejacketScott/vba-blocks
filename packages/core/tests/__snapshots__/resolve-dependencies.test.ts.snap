// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolves simple dependency tree 1`] = `
Map {
  "a" => Object {
    "name": "a",
    "needs": Array [
      "^1.0.0",
    ],
    "required": true,
    "versions": Array [
      Object {
        "dependencies": Array [
          Object {
            "name": "d",
            "version": "^2.0.0",
          },
        ],
        "id": "a@1.2.0",
        "name": "a",
        "version": "1.2.0",
      },
      Object {
        "dependencies": Array [
          Object {
            "name": "d",
            "version": "^1.2.0",
          },
        ],
        "id": "a@1.1.0",
        "name": "a",
        "version": "1.1.0",
      },
      Object {
        "dependencies": Array [
          Object {
            "name": "d",
            "version": "^1.0.0",
          },
        ],
        "id": "a@1.0.0",
        "name": "a",
        "version": "1.0.0",
      },
    ],
  },
  "d" => Object {
    "name": "d",
    "needs": Array [
      "^1.0.0",
      "^1.2.0",
      "^2.0.0",
    ],
    "required": false,
    "versions": Array [
      Object {
        "dependencies": Array [
          Object {
            "name": "f",
            "version": "2.0.0",
          },
        ],
        "id": "d@2.0.0",
        "name": "d",
        "version": "2.0.0",
      },
      Object {
        "dependencies": Array [
          Object {
            "name": "f",
            "version": "1.0.0",
          },
        ],
        "id": "d@1.2.0",
        "name": "d",
        "version": "1.2.0",
      },
      Object {
        "dependencies": Array [
          Object {
            "name": "f",
            "version": "1.0.0",
          },
        ],
        "id": "d@1.1.0",
        "name": "d",
        "version": "1.1.0",
      },
      Object {
        "dependencies": Array [
          Object {
            "name": "f",
            "version": "1.0.0",
          },
        ],
        "id": "d@1.0.0",
        "name": "d",
        "version": "1.0.0",
      },
    ],
  },
  "f" => Object {
    "name": "f",
    "needs": Array [
      "1.0.0",
      "2.0.0",
    ],
    "required": false,
    "versions": Array [
      Object {
        "dependencies": Array [
          Object {
            "name": "g",
            "version": "^1",
          },
        ],
        "id": "f@2.0.0",
        "name": "f",
        "version": "2.0.0",
      },
      Object {
        "dependencies": Array [
          Object {
            "name": "g",
            "version": "^1",
          },
        ],
        "id": "f@1.0.0",
        "name": "f",
        "version": "1.0.0",
      },
    ],
  },
  "g" => Object {
    "name": "g",
    "needs": Array [
      "^1",
    ],
    "required": false,
    "versions": Array [
      Object {
        "dependencies": Array [],
        "id": "g@1.0.0",
        "name": "g",
        "version": "1.0.0",
      },
    ],
  },
}
`;
